src/goto-programs/goto_program.cpp

```diff
@@ -532,10 +533,33 @@ std::string as_string(
 /// order they appear in the goto program.
 void goto_programt::compute_loop_numbers()
 {
+  // This deals with nested loop number assignment by treating the queue
+  // as a loop stack, and it'll hold whatever number is currently considered
+  // active.
   unsigned nr=0;
+  std::queue<int> active_loops;
   for(auto &i : instructions)
+  {
+    bool loop_target = false;
+    for(auto &inc_edge : i.incoming_edges)
+      loop_target = loop_target || inc_edge->is_backwards_goto();
+
+    if (loop_target)
+    {
+      active_loops.push(nr++);
+    }
+
+    // Assign the loop number to every instruction associated with that loop.
+    if (!active_loops.empty())
+    {
+      i.loop_number = active_loops.back();
+    }
+
     if(i.is_backwards_goto())
```

Comment:
This is already the case in goto-symex, but suggest commenting explicitly that this notion of loop is inadequate, backward-gotos are not necessarily in 1-to-1 correspondence with loops.